### PR TITLE
SDI-1995 Move physical attributes to own package and replace duplicate name prefixes

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/EventListener.kt [hunk 1]
```diff
@@ -10,14 +10,16 @@ import kotlinx.coroutines.Dispatchers
...
 import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.listeners.SQSMessage
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.prisonperson.physicalattributes.SynchronisationService
 import java.util.concurrent.CompletableFuture
 
-@Service
-class PrisonPersonEventListener(
-  private val prisonPersonService: PrisonPersonSynchronisationService,
+@Service("prisonPersonEventListener")
+class EventListener(
+  @Qualifier("physicalAttributesSynchronisationService") private val physicalAttributesSyncService: SynchronisationService,
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/EventListener.kt:22]: I've removed the `PrisonPerson` prefix from all classes as they make it hard on the eyes. So there is lots of renaming / qualifying beans so they don't clash with other packages.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/MigrationResource.kt [hunk 2]
```diff
@@ -20,38 +19,13 @@ import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.service.HistoryFi
...
 @RequestMapping("/migrate/prisonperson", produces = [MediaType.APPLICATION_JSON_VALUE])
-class PrisonPersonMigrationResource(
-  private val prisonPersonMigrationService: PrisonPersonMigrationService,
+@Tag(name = "prison-person-migration-resource")
+class MigrationResource(
+  private val migrationService: MigrationService,
   private val migrationHistoryService: MigrationHistoryService,
 ) {
-  @PreAuthorize("hasRole('ROLE_MIGRATE_PRISONPERSON')")
-  @PostMapping("/physical-attributes")
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/MigrationResource.kt:30]: Anything specific to physical attributes has been moved up into the physical attributes package.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/MigrationService.kt [hunk 3]
```diff
@@ -21,51 +23,57 @@ import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.service.Migration
...
+  @Qualifier("prisonPersonMappingApiService") private val mappingApiService: MappingApiService,
   migrationHistoryService: MigrationHistoryService,
   telemetryClient: TelemetryClient,
   auditService: AuditService,
   @Value("\${page.size:1000}") pageSize: Long,
   @Value("\${complete-check.delay-seconds}") completeCheckDelaySeconds: Int,
   @Value("\${complete-check.count}") completeCheckCount: Int,
-) : MigrationService<PrisonPersonMigrationFilter, PrisonPersonMigrationRequest, PrisonerPhysicalAttributesResponse, PrisonPersonMigrationMappingRequest>(
+  private val entityMigrators: List<EntityMigrator<*, *>>,
+) : MigrationService<MigrationFilter, MigrationRequest, PrisonerPhysicalAttributesResponse, PrisonPersonMigrationMappingRequest>(
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/MigrationService.kt:38]: The third generic type relates still relates to physical attributes - not sure why this is required, will try and refactor ir out next.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/MigrationService.kt [hunk 4]
```diff
@@ -21,51 +23,57 @@ import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.service.Migration
...
   pageSize = pageSize,
   completeCheckDelaySeconds = completeCheckDelaySeconds,
   completeCheckCount = completeCheckCount,
 ) {
-  abstract suspend fun migrateEntity(offenderNo: String): List<Long>
+  private suspend fun migrateEntity(
+    offenderNo: String,
+    migrationType: PrisonPersonMigrationMappingRequest.MigrationType,
+  ): List<Long> = entityMigrators.find { it.supportsType(migrationType.javaClass) }?.migrateEntity(offenderNo)
+    ?: throw IllegalArgumentException("No entity migrator found for migration type $migrationType")
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/MigrationService.kt:53]: For the type specific entity migrations I've had to resort to the visitor pattern based on the MigrationType. It got very messy with generics so dropped that, but there might be a better way than this 🤔

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/MigrationService.kt [hunk 5]
```diff
@@ -21,51 +23,56 @@ import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.service.Migration
...
   migrationType = MigrationType.PRISONPERSON,
   pageSize = pageSize,
   completeCheckDelaySeconds = completeCheckDelaySeconds,
   completeCheckCount = completeCheckCount,
 ) {
-  abstract suspend fun migrateEntity(offenderNo: String): List<Long>
+  private suspend fun migrateEntity(
+    offenderNo: String,
+    migrationType: PrisonPersonMigrationMappingRequest.MigrationType,
+  ): DpsResponse = entityMigratorService.migrator(migrationType).migrate(offenderNo)
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/MigrationService.kt:52]: Adds support for different migration types, each of which will have their own EntityMigrator

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/PrisonPersonEntityMigrator.kt [hunk 6]
```diff
@@ -1,12 +0,0 @@
...
-
-interface PrisonPersonEntityMigrator<NOMIS_RESPONSE, DPS_REQUEST> {
-  suspend fun migrateEntity(offenderNo: String): List<Long> =
-    getNomisEntity(offenderNo)
-      .toDpsMigrationRequests()
-      .migrate(offenderNo)
-
-  suspend fun getNomisEntity(offenderNo: String): NOMIS_RESPONSE
-  suspend fun NOMIS_RESPONSE.toDpsMigrationRequests(): List<DPS_REQUEST>
-  suspend fun List<DPS_REQUEST>.migrate(offenderNo: String): List<Long>
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/PrisonPersonEntityMigrator.kt:11]: Got rid of the generic entity migrator and replaced above with slimmed down interface that doesn't expose the internal types which are just details:
```
interface EntityMigrator {
  suspend fun migrate(offenderNo: String): DpsResponse
}
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/EventListener.kt (hunk 1 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/EventListener.kt:22]: Although if everyone did that Command-O would be pretty useless in IntelliJ right?

**Review by andymarke — APPROVED** [review]
lgtm

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/EventListener.kt (hunk 1 above)
Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/EventListener.kt:22]: I'm going to have lots of these for my various migrations. Let's see how it goes. I despise prefixing class names with the package name 😆

**Pull request merged**